src/shims/time.rs

```diff
@@ -1,8 +1,145 @@
-use std::time::{Duration, Instant, SystemTime};
+use std::sync::atomic::AtomicU64;
+use std::time::{Duration, Instant as StdInstant, SystemTime};
+
+use rustc_data_structures::sync::Ordering;
 
 use crate::concurrency::thread::Time;
 use crate::*;
 
+/// When using a virtual clock, this defines how many nanoseconds do we pretend
+/// are passing for each basic block.
+const NANOSECOND_PER_BASIC_BLOCK: u64 = 10;
+
+#[derive(Debug)]
+pub struct Instant {
+    kind: InstantKind,
+}
+
+#[derive(Debug)]
+enum InstantKind {
+    Host(StdInstant),
+    Virtual { nanoseconds: u64 },
+}
+
+/// A monotone clock used for `Instant` simulation.
+#[derive(Debug)]
+pub struct Clock {
+    kind: ClockKind,
+}
+
+#[derive(Debug)]
+enum ClockKind {
+    Host {
+        /// The "time anchor" for this machine's monotone clock.
+        time_anchor: StdInstant,
+    },
+    Virtual {
+        /// The "current virtual time".
+        nanoseconds: AtomicU64,
+    },
+}
+
+impl Clock {
+    /// Create a new clock based on the availability of communication with the host.
+    pub fn new(communicate: bool) -> Self {
+        let kind = if communicate {
+            ClockKind::Host { time_anchor: StdInstant::now() }
+        } else {
+            ClockKind::Virtual { nanoseconds: 0.into() }
+        };
+
+        Self { kind }
+    }
+
+    /// Get the current time relative to this clock.
+    pub fn get(&self) -> Duration {
```

Comment:
Now we have `get` and `get_time_relative/absolute` and they have quite different types... that's somewhat confusing. Hm...